security/advancedtls/advancedtls_integration_test.go

```diff
@@ -408,3 +411,272 @@ func (s) TestEnd2End(t *testing.T) {
 		})
 	}
 }
+
+// In order to test advanced TLS provider features, we used temporary files to
+// hold credential data, and copy the contents under testdata/ to these tmp
+// files.
+// Initially, we establish a good connection with providers watching contents
+// from tmp files.
+// Next, we change the identity certs that IdentityProvider is watching. Since
+// the identity key is not changed, the IdentityProvider should ignore the
+// update, and the connection should still be good.
+// Then the the identity key is changed. This time IdentityProvider should pick
+// up the update, and the connection should fail, due to the trust certs on the
+// other side is not changed.
+// Finally, the trust certs that other-side's RootProvider is watching get
+// changed. The connection should go back to normal again.
+func (s) TestPEMFileProviderEnd2End(t *testing.T) {
+	// Create temp files that are used to hold credentials.
+	clientCertTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(clientCertTmp.Name())
+	clientKeyTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(clientKeyTmp.Name())
+	clientTrustTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(clientTrustTmp.Name())
+	serverCertTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(serverCertTmp.Name())
+	serverKeyTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(serverKeyTmp.Name())
+	serverTrustTmp, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Fatalf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(serverTrustTmp.Name())
+	for _, test := range []struct {
+		desc                string
+		certUpdateFunc      func()
+		keyUpdateFunc       func()
+		trustCertUpdateFunc func()
+	}{
+		{
+			desc: "TestClientPeerCertProviderServerTrustCertProvider",
```

Comment:
These sub test names are not very readable. They do not clearly indicate what is being done in the sub test. Could you please come up with more meaningful test descriptions.